proto.c

```diff
@@ -744,6 +744,70 @@ static void handle_tcp_packet(char *packet_buff, int packet_buff_len,
 	return;
 }
 
+static void handle_icmp_packet(char *packet_buff, int packet_buff_len,
+			      struct node *node)
+{
+	struct icmphdr *icmphdr;
+	struct ether_header *out_ethhdr;
+	struct iphdr *out_iphdr;
+	struct icmphdr *out_icmphdr;
+	size_t pos;
+
+	if (!len_check(packet_buff_len, sizeof(struct icmphdr), "ICMP"))
+		goto out;
+
+	icmphdr = (struct icmphdr *)packet_buff;
+
+	/* not echo request */
+	if (icmphdr->type != 8)
+		goto out;
+
+	if (icmphdr->code != 0)
+		goto out;
+
+	pos = 0;
+	out_ethhdr = (struct ether_header *)&out_packet_buff[pos];
```

Comment:
Why redeclaring all the out_* fields when they are already available as global variable? Any reason for that?